util/tracing: remove noop span

xinhaoz opened the pull request:
This commit removes the concept of a low overhead 'noop span' as
this is obsolete. The noop span was primarily used to prevent
unnnecessary allocations and operations when tracing was off,
while enabling access to its wrapped tracer to start spans later on.

Summary:
- Where we were creating noop spans, we are now using `nil`
spans - this includes places where we used sterile noop spans.
- All checks for `span.IsNoop` have been removed where there was
already a `nil` check, or have been replaced with `nil` checks.
- Places where we expected to have at least a noop  span from the
context with a valid tracer have been replaced to retrieve the
tracer from a different source, if necessary, if the span is nil.

Part of: https://github.com/cockroachdb/cockroach/issues/135803

Release note: None

xinhaoz committed:
util/tracing: remove noop spans

This commit removes the concept of a low overhead 'noop span' as
this is obsolete. The noop span was primarily used to prevent
unnnecessary allocations and operations when tracing was off,
while enabling access to its wrapped tracer to start spans later on.

Summary:
- Where we were creating noop spans, we are now using `nil`
spans - this includes places where we used sterile noop spans.
- All checks for `span.IsNoop` have been removed where there was
already a `nil` check, or have been replaced with `nil` checks.
- Places where we expected to have at least a noop  span from the
context with a valid tracer have been replaced to retrieve the
tracer from a different source, if necessary, if the span is nil.

Part of: #135803

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139352)
<!-- Reviewable:end -->

xinhaoz commented:
```

name                                                                            old time/op    new time/op    delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                   62.6ns ±13%    13.7ns ± 4%   -78.15%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                    20.6ns ±82%    13.7ns ±18%   -33.44%  (p=0.000 n=9+10)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                          643ns ± 4%     610ns ±12%    -5.20%  (p=0.012 n=8+10)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                       351ns ± 2%     343ns ± 7%    -2.21%  (p=0.011 n=8+9)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                    305ns ± 3%     302ns ± 2%      ~     (p=0.098 n=9+9)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                     597ns ± 4%     596ns ± 4%      ~     (p=0.949 n=9+9)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                             322ns ± 2%     320ns ± 5%      ~     (p=0.661 n=9+10)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                              607ns ± 4%     608ns ± 4%      ~     (p=0.666 n=9+9)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                         323ns ± 3%     317ns ± 3%      ~     (p=0.113 n=9+9)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                        659ns ± 3%     649ns ± 8%      ~     (p=0.200 n=9+8)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10       478ns ± 2%     475ns ± 4%      ~     (p=0.393 n=10+10)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10       1.12µs ± 6%    1.10µs ± 4%      ~     (p=0.225 n=10+10)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10     366ns ± 4%     360ns ± 2%      ~     (p=0.075 n=10+10)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10      668ns ± 5%     669ns ± 5%      ~     (p=1.000 n=9+9)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                             772ns ± 6%     763ns ± 3%      ~     (p=0.604 n=10+9)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                              841ns ± 4%     828ns ± 3%      ~     (p=0.252 n=9+7)

name                                                                            old alloc/op   new alloc/op   delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                     160B ± 0%        0B       -100.00%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                      160B ± 0%        0B       -100.00%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                     160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                      160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                              160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                               160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                          160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                           160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                        160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                         160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10        208B ± 0%      208B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10         208B ± 0%      208B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10      160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10       160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                              160B ± 0%      160B ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                               163B ± 0%      163B ± 0%      ~     (all equal)

name                                                                            old allocs/op  new allocs/op  delta
Tracer_StartSpanCtx/opts=none/parallel=false-10                                     1.00 ± 0%      0.00       -100.00%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                      1.00 ± 0%      0.00       -100.00%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=real/parallel=false-10                                     1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real/parallel=true-10                                      1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=false-10                              1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,logtag/parallel=true-10                               1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=false-10                          1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent/parallel=true-10                           1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=false-10                        1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent/parallel=true-10                         1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=false-10        2.00 ± 0%      2.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,autoparent,withEventListener/parallel=true-10         2.00 ± 0%      2.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=false-10      1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,manualparent,withEventListener/parallel=true-10       1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=false-10                              1.00 ± 0%      1.00 ± 0%      ~     (all equal)
Tracer_StartSpanCtx/opts=real,regexp/parallel=true-10                               1.00 ± 0%      1.00 ± 0%      ~     (all equal)
```

xinhaoz commented:
This is ready for a look, please let me know if I missed or have misunderstood something.

dhartunian reviewed (commented):
Nice work going through this. No major concerns. Just questions about whether some behavior can be removed entirely.

> Reviewed 21 of 22 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-:-OHJT-6a65RGfOIuwpbr:b-g25qkf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arjunmahishi, @kev-cao, @rytaft, and @xinhaoz)

___
*[`pkg/util/tracing/tracer_test.go` line 785 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJStu3-3JEYka1iEe8:-OHJStu3-3JEYka1iEe9:b-v5wvig) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/tracer_test.go#L785)):*
> ```Go
> 		noopParent    bool
> 	}
> 	// TODO (xinhaoz): delete
> ```

reminder :)
___
*[`pkg/util/tracing/context.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJRHQMDVBYjf2ifv9C:-OHJRHQMDVBYjf2ifv9D:b-742qt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/context.go#L59)):*
> ```Go
> 		}
> 	}
> 	return ctxutil.WithFastValue(ctx, activeSpanKey, sp), sp
> ```

This function is now a bit confusing. `nil` returning from `SpanFromContext` doesn't clarify whether we've set the context value or not. It could be set to `nil` or it could be unset.

I wonder if we should just return when `sp == nil`.
___
*[`pkg/util/tracing/tracer.go` line 1126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJQchW8FmNo4FOsOa-:-OHJQchW8FmNo4FOsOa0:bpen5a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/tracer.go#L1126)):*
> ```Go
> 		forceVerbose := t.forceOpNameVerbose(opName)
> 		if shouldBeNilSpan && !forceVerbose && !opts.Sterile {
> 			return maybeWrapCtx(ctx, nil)
> ```

should this just do nothing and return? wrapping with a nil span is a no-op, right?
___
*[`pkg/util/tracing/tracer.go` line 1178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJQy1SDZVUFsQdvMCv:-OHJQy1SDZVUFsQdvMCw:b-o4n2pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/tracer.go#L1178)):*
> ```Go
> 			return maybeWrapCtx(ctx, nil)
> 		}
> 		return maybeWrapCtx(ctx, nil)
> ```

same thing here, should these calls just be removed?
___
*[`pkg/util/tracing/tracer.go` line 1183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJQs2432WPBhOwF-3c:-OHJQs2432WPBhOwF-3d:b1ihowj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f2c9b09da6f7fee1446859b41a55be33ae1fd07/pkg/util/tracing/tracer.go#L1183)):*
> ```Go
> 		if opts.Parent.IsNoop() {
> 			// If the parent is a no-op, we'll create a root span.
> 			opts.Parent = spanRef{}
> ```

should this be `opts.Parent == nil`? If the parent is empty then we should be creating a root span.
___
*[`pkg/backup/backup_processor.go` line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHF6Twm5H8HFtHyup4K:-OHF6Twm5H8HFtHyup4L:b-sxpsw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/backup/backup_processor.go#L567)):*
> ```Go
> 								tracer := sp.Tracer()
> 								if tracer == nil {
> 									tracer = flowCtx.Cfg.Tracer
> ```

is this guaranteed to be non-nil?


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139352)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arjunmahishi, @dhartunian, @kev-cao, and @rytaft)

___
*[`pkg/backup/backup_processor.go` line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHF6Twm5H8HFtHyup4K:-OHK3iR_EvQJ9GUEj264:b-gvryd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/backup/backup_processor.go#L567)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

is this guaranteed to be non-nil?
</blockquote></details>

As far as I could tell from tracing it up the call stack, yeah it should be set. Perhaps still better not to make that assumption though. Added conditional so that we only trace if we have a valid tracer.
___
*[`pkg/util/tracing/context.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJRHQMDVBYjf2ifv9C:-OHK9Lde5JBxhiione85:bm3rgw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/context.go#L59)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This function is now a bit confusing. `nil` returning from `SpanFromContext` doesn't clarify whether we've set the context value or not. It could be set to `nil` or it could be unset.

I wonder if we should just return when `sp == nil`.
</blockquote></details>

According to the comment (which I just put back), callers can pass nil to remove a span from the context, so i don't think we should return if the passed span is nil.
___
*[`pkg/util/tracing/tracer.go` line 1126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJQchW8FmNo4FOsOa-:-OHKAUUf8KsN2rMCXGzZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/tracer.go#L1126)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should this just do nothing and return? wrapping with a nil span is a no-op, right?
</blockquote></details>

Done.
___
*[`pkg/util/tracing/tracer.go` line 1178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJQy1SDZVUFsQdvMCv:-OHK7fgK8BIIhwIFgWjG:b-g1le1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/tracer.go#L1178)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

same thing here, should these calls just be removed?
</blockquote></details>

Ah yup nice catch - done.
___
*[`pkg/util/tracing/tracer.go` line 1183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJQs2432WPBhOwF-3c:-OHJblYv9651HAfJDeCJ:bxp4mp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f2c9b09da6f7fee1446859b41a55be33ae1fd07/pkg/util/tracing/tracer.go#L1183)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should this be `opts.Parent == nil`? If the parent is empty then we should be creating a root span.
</blockquote></details>

Parent being nil should lead to a root span, see comment here: https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/span_options.go#L166-L171.

Parent field is always non-nil as it's actually a concrete `spanRef` which wraps a `*Span`.
___
*[`pkg/util/tracing/tracer_test.go` line 785 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJStu3-3JEYka1iEe8:-OHK9C9f4JuZyGbQvV4E:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/tracer_test.go#L785)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

reminder :)
</blockquote></details>

Done, thanks!
___
*[`pkg/util/tracing/tracer.go` line 1157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJdf920W7NgF560rw3:-OHJdf920W7NgF560rw4:bhe6gji) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f2c9b09da6f7fee1446859b41a55be33ae1fd07/pkg/util/tracing/tracer.go#L1157)):*
> ```Go
> 		}
> 
> 		if !opts.RemoteParent.Empty() {
> ```

Huh, I'm not sure why I removed this on first pass. I put it back.


<!-- Sent from Reviewable.io -->

xinhaoz committed:
util/tracing: remove isNoop checks from `spanInner`

In the previous commit we removed the top level noop
span. This commit removes `isNoop` checks from the
`spanInner` struct, which is the underlying span struct
wrapped by the top level span. This check returned false if
all span types in the struct are nil.

Note that we should no  longer be creating `spanInner` structs
without defining one of `crdb`, `otelSpan`, or `netTr`, and doing
so will lead to a fatal error.  Comments have been updated to
highlight this.

This commit also removes the `opts=none` benchmark in
`BenchmarkTracer_StartSpanCtx` as regressions are more likely
to occur now with unrelated changes.

Epic: none
Part of: #135803

Release note: None

xinhaoz commented:
@dhartunian I updated with another commit also removing isNoop checks from the inner span to clean things up further.

xinhaoz commented on pkg/util/tracing/tracer.go:
```diff
@@ -1127,10 +1120,10 @@ func (t *Tracer) startSpanFast(
-		if shouldBeNoopSpan && !forceVerbose && !opts.Sterile {
-			return maybeWrapCtx(ctx, t.noopSpan)
+		if shouldBeNilSpan && !forceVerbose && !opts.Sterile {
+			return maybeWrapCtx(ctx, nil)
```
@dhartunian  I ended up keeping these instead of just returning `ctx, nil`. It lead to some failied tests without. I'm guessing there are cases where ctx has a non nil span and the caller of startspan is expecting to clear that (or get its own span).

dhartunian reviewed (commented):
> Reviewed 3 of 4 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139352)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arjunmahishi and @xinhaoz)

___
*[`pkg/backup/backup_processor.go` line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHF6Twm5H8HFtHyup4K:-OHcP4Re9cpFQBl0C7zM:bkh9r35) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/backup/backup_processor.go#L567)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

As far as I could tell from tracing it up the call stack, yeah it should be set. Perhaps still better not to make that assumption though. Added conditional so that we only trace if we have a valid tracer.
</blockquote></details>

I think `StartSpanCtx` should handle the case where the tracer is nil given the fact that `noopSpan` doesn't exist anymore. The code here seems brittle and I don't think users of the tracing lib should be expected to do so much checking.
___
*[`pkg/util/tracing/context.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHJRHQMDVBYjf2ifv9C:-OHcQwya0XWYjYIJDCv7:bmdn7rw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/util/tracing/context.go#L59)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

According to the comment (which I just put back), callers can pass nil to remove a span from the context, so i don't think we should return if the passed span is nil.
</blockquote></details>

Ack. Thanks for the explanation.


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139352)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arjunmahishi and @dhartunian)

___
*[`pkg/backup/backup_processor.go` line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-OHF6Twm5H8HFtHyup4K:-OHcXb9eB85beh3YWd7l:bxrm5k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3e0a9069e0e7e6f3e3469f6d2604a53d02a94d/pkg/backup/backup_processor.go#L567)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think `StartSpanCtx` should handle the case where the tracer is nil given the fact that `noopSpan` doesn't exist anymore. The code here seems brittle and I don't think users of the tracing lib should be expected to do so much checking.
</blockquote></details>

Makes sense to me - done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139352#-:-OHdNdqR5V593MR1QiHM:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arjunmahishi and @xinhaoz)



<!-- Sent from Reviewable.io -->

mgartner commented:
Nice work! These two benchmarks are now so fast, that changes to completely unrelated code that change how instructions are laid out in memory or the alter the compiler's inlining decisions are likely to cause a "regressions" which are reported in #perf-ops. It's probably safe to remove these two cases entirely.

```
Tracer_StartSpanCtx/opts=none/parallel=false-10                                   62.6ns ±13%    13.7ns ± 4%   -78.15%  (p=0.000 n=10+10)
Tracer_StartSpanCtx/opts=none/parallel=true-10                                    20.6ns ±82%    13.7ns ±18%   -33.44%  (p=0.000 n=9+10)
```

xinhaoz commented:
@mgartner Thanks for the call out, I've removed the `opts=none` benchmark from `BenchmarkTracer_StartSpanCtx`.

xinhaoz commented on pkg/backup/backup_processor.go:
@kyle-a-wong We've modified `StartSpanCtx` to handle  a nil tracer, see `tracer.go` diff.

xinhaoz commented:
tftr!
bors r+

xinhaoz commented:
bors r+

tbg commented:
Nice, thanks for pushing on this!

> The self-hosted runner: basic-runner-group-20250129-162228-048319 lost communication with the server. Verify the machine is running and has a healthy network connection. Anything in your workflow that terminates the runner process, starves it for CPU/Memory, or blocks its network access can cause this error.

I'll poke this again.

bors r+
